src/main/java/org/apache/commons/text/CasedString.java

```diff
@@ -0,0 +1,188 @@
+
+        /** test for split position character. */
+        private final Predicate<Character> splitter;
+        /** if {@code true} split position character will be preserved in following segment. */
+        private final boolean preserveSplit;
+        /** a function to joining the segments into this case type. */
+        private final Function<String[], String> joiner;
+
+        /**
+         * Defines a String Case.
+         * @param splitter The predicate that determines when a new word in the cased string begins.
+         * @param preserveSplit if {@code true} the character that the splitter detected is preserved as the first character of the new word.
+         * @param joiner The function to merge a list of strings into the cased String.
+         */
+        StringCase(final Predicate<Character> splitter, final boolean preserveSplit, final Function<String[], String> joiner) {
+            this.splitter = splitter;
+            this.preserveSplit = preserveSplit;
+            this.joiner = joiner;
+        }
+
+        /**
+         * Creates a cased string from a collection of segments.
+         * @param segments the segments to create the CasedString from.
+         * @return a CasedString
+         */
+        public String assemble(String[] segments) {
+            return segments.length == 0 ? null : this.joiner.apply(segments);
+        }
+
+        /**
+         * Returns an array of each of the segments in this CasedString.  Segments are defined as the strings between
+         * the separators in the CasedString.  for the CAMEL case the segments are determined by the presence of a capital letter.
+         * @return the array of Strings that are segments of the cased string.
+         */
+        public String[] getSegments(String string) {
+            if (string == null) {
+                return new String[0];
+            }
+            if (string.isEmpty()) {
+                return new String[]{""};
+            }
+            List<String> lst = new ArrayList<>();
+            StringBuilder sb = new StringBuilder();
+            for (char c : string.toCharArray()) {
+                if (splitter.test(c)) {
+                    if (sb.length() > 0) {
+                        lst.add(sb.toString());
+                        sb.setLength(0);
+                    }
+                    if (preserveSplit) {
+                        sb.append(c);
+                    }
+                } else {
+                    sb.append(c);
+                }
+            }
+            if (sb.length() > 0) {
+                lst.add(sb.toString());
+            }
+            return lst.toArray(new String[0]);
+        }
+    }
+
+    /**
+     * A representation of a cased string and the identified case of that string.
+     * @param stringCase The {@code StringCase} that the {@code string} argument is in.
+     * @param string The string.
+     */
+    public CasedString(StringCase stringCase, String string) {
```

Comment:
This constructor should not let me construct illegal instances, for example " a b c " should not be an allowed camel case string (or a kebab, or a snake string). Otherwise, it's just GIGO.

Reply:
That's not acceptable IMO, when I create a BigInteger or a Duration, I get an exception when I feed the methods garbage. It should be the same here, fail-fast is always better in this case, otherwise, I could get all sorts of junk down the line. 

> I expect that you know what the string looks like that you are starting with. 

This will not only be for user input, where typos could still happen of course. 

When you start with an Open API or XML Schema, you hope the input is well-formed and valid, but even if so, the contents may be junky.